pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -889,31 +890,70 @@ protected void doTakeSnapshot() {
+        _logger.warn("Could not get segmentLock to take snapshot for segment: {}, skipping", segmentName);
+        skippedSegment = true;
+        continue;
+      }
       try {
         ImmutableSegmentImpl immutableSegment = (ImmutableSegmentImpl) segment;
         if (!immutableSegment.hasValidDocIdsSnapshotFile()) {
           segmentsWithoutSnapshot.add(immutableSegment);
           continue;
         }
         immutableSegment.persistValidDocIdsSnapshot();
+        _updatedSegmentsSinceLastSnapshot.remove(segment);
         numImmutableSegments++;
         numPrimaryKeysInSnapshot += immutableSegment.getValidDocIds().getMutableRoaringBitmap().getCardinality();
       } catch (Exception e) {
-        _logger.warn("Caught exception while taking snapshot for segment: {}, skipping", segment.getSegmentName(), e);
-        Utils.rethrowException(e);
+        _logger.warn("Caught exception while taking snapshot for segment: {}, skipping", segmentName, e);
+        skippedSegment = true;
+      } finally {
+        segmentLock.unlock();
       }
     }
-    for (ImmutableSegmentImpl segment : segmentsWithoutSnapshot) {
-      try {
-        segment.persistValidDocIdsSnapshot();
-        numImmutableSegments++;
-        numPrimaryKeysInSnapshot += segment.getValidDocIds().getMutableRoaringBitmap().getCardinality();
-      } catch (Exception e) {
-        _logger.warn("Caught exception while taking snapshot for segment: {}, skipping", segment.getSegmentName(), e);
-        Utils.rethrowException(e);
+    // If we skipped any segments with existing snapshot files in the for-loop above, then we should not take snapshots
+    // for segments w/o snapshot files yet, i.e. skip the next for-loop to not add new snapshot files on disk. This
+    // ensures that the validDocIds snapshots kept on disk are all disjoint as to the PKs tracked by them, in order for
+    // segment preloading to work correctly.
+    if (!skippedSegment) {
+      for (ImmutableSegmentImpl segment : segmentsWithoutSnapshot) {
+        String segmentName = segment.getSegmentName();
+        Lock segmentLock = tableDataManager.getSegmentLock(segmentName);
+        boolean locked = segmentLock.tryLock();
+        if (!locked) {
+          _logger.warn("Could not get segmentLock to take snapshot for segment: {} w/o snapshot, skipping",
+              segmentName);
+          continue;
+        }
+        try {
+          segment.persistValidDocIdsSnapshot();
+          _updatedSegmentsSinceLastSnapshot.remove(segment);
+          numImmutableSegments++;
+          numPrimaryKeysInSnapshot += segment.getValidDocIds().getMutableRoaringBitmap().getCardinality();
+        } catch (Exception e) {
+          _logger.warn("Caught exception while taking snapshot for segment: {} w/o snapshot, skipping", segmentName, e);
+        } finally {
+          segmentLock.unlock();
+        }
       }
     }
-    _updatedSegmentsSinceLastSnapshot.clear();
```

Comment:
Is there guarantee that all segments are cleared here? Seems we don't remove segments when they are deleted.

Reply:
good catch. To be simple, I'll do retainAll(_trackedSegments) here to avoid keeping removed segments around